File: luaprobe.cpp
// Cross-platform Lua-math divergence probe.
// LuaJIT's math.sin/cos/tan/asin/acos/tanh/log wrap the C library (libm); the ^ operator
// uses pow. This dumps raw IEEE-754 bits of each over a dense domain sweep so arm64 (Apple)
// and Linux (glibc) outputs can be byte-diffed per op. Build: clang++/g++ -O2 -ffp-contract=off
#include <cstdio>
#include <cmath>
#include <cstdint>
#include <cstring>

static uint64_t B(double d) { uint64_t u; std::memcpy(&u, &d, 8); return u; }

static void sweep(FILE* f, const char* name, double (*fn)(double), double lo, double hi, int n) {
	for (int i = 0; i <= n; ++i) {
		double x = lo + (hi - lo) * (static_cast<double>(i) / n);
		double y = fn(x);
		std::fprintf(f, "%s,%016llx,%016llx\n", name,
		             static_cast<unsigned long long>(B(x)), static_cast<unsigned long long>(B(y)));
	}
}

static void sweepPow(FILE* f, const char* name, double e, double lo, double hi, int n) {
	for (int i = 0; i <= n; ++i) {
		double b = lo + (hi - lo) * (static_cast<double>(i) / n);
		double y = std::pow(b, e);
		std::fprintf(f, "%s,%016llx:%016llx,%016llx\n", name,
		             static_cast<unsigned long long>(B(b)), static_cast<unsigned long long>(B(e)),
		             static_cast<unsigned long long>(B(y)));
	}
}

int main(int argc, char** argv) {
	const char* out = argc > 1 ? argv[1] : "luaprobe.csv";
	FILE* f = std::fopen(out, "w");
	const int N = 4000;
	sweep(f, "sin",  std::sin,  -12.566370614359172, 12.566370614359172, N); // [-4pi,4pi]
	sweep(f, "cos",  std::cos,  -12.566370614359172, 12.566370614359172, N);
	sweep(f, "tan",  std::tan,  -1.5, 1.5, N);
	sweep(f, "asin", std::asin, -1.0, 1.0, N);
	sweep(f, "acos", std::acos, -1.0, 1.0, N);
	sweep(f, "tanh", std::tanh, -10.0, 10.0, N);
	sweep(f, "sinh", std::sinh, -10.0, 10.0, N);
	sweep(f, "cosh", std::cosh, -10.0, 10.0, N);
	sweep(f, "log",  std::log,  1e-12, 1000.0, N);
	sweep(f, "exp",  std::exp,  -700.0, 700.0, N);
	sweepPow(f, "pow_int3", 3.0, -5.0, 5.0, N);   // x^3 (the SharedBehaviors aimSkill call)
	sweepPow(f, "pow_0.37", 0.37, 0.001, 10.0, N);
	sweepPow(f, "pow_2.5",  2.5, 0.001, 10.0, N);
	std::fclose(f);
	std::printf("wrote %s\n", out);
	return 0;
}

File: luaprobe.lua
-- Cross-arch LuaJIT transcendental divergence probe (round 16).
-- Dumps op,index,arg-bits,result-bits (raw 8-byte LE hex of the double) for a
-- dense sweep of each op over its domain. Align the cross-platform diff by
-- (op,index): both platforms use identical integer index -> identical intended
-- arg, so any result-bit difference is a genuine libm/op divergence.
local ffi = require("ffi")
local buf = ffi.new("double[1]")
local function bits(x)
	buf[0] = x
	local p = ffi.cast("uint8_t*", buf)
	return string.format("%02x%02x%02x%02x%02x%02x%02x%02x",
		p[0], p[1], p[2], p[3], p[4], p[5], p[6], p[7])
end

-- name, fn, domain lo, hi  (domains kept strictly valid -> no NaN, JIT-friendly)
local ops = {
	{ "sin",      function(x) return math.sin(x) end,  -12.5, 12.5 },
	{ "cos",      function(x) return math.cos(x) end,  -12.5, 12.5 },
	{ "tan",      function(x) return math.tan(x) end,   -1.5,  1.5 },
	{ "asin",     function(x) return math.asin(x) end,  -1.0,  1.0 },
	{ "acos",     function(x) return math.acos(x) end,  -1.0,  1.0 },
	{ "atan",     function(x) return math.atan(x) end, -12.5, 12.5 },
	{ "tanh",     function(x) return math.tanh(x) end,  -8.0,  8.0 },
	{ "exp",      function(x) return math.exp(x) end,   -8.0,  8.0 },
	{ "log",      function(x) return math.log(x) end,  1e-6,  50.0 },
	{ "sqrt",     function(x) return math.sqrt(x) end,   0.0, 100.0 },
	{ "powfn2.5", function(x) return math.pow(x, 2.5) end, 0.0, 10.0 },
	{ "pow_op25", function(x) return x ^ 2.5 end,         0.0, 10.0 }, -- ^ non-integer
	{ "pow_op3",  function(x) return x ^ 3 end,         -10.0, 10.0 }, -- ^ integer exponent
}

local N = 2000
local out = assert(io.open(arg[1] or "luaprobe.txt", "w"))
for _, op in ipairs(ops) do
	local name, f, lo, hi = op[1], op[2], op[3], op[4]
	local step = (hi - lo) / N
	for i = 0, N do
		local x = lo + step * i
		local y = f(x)
		if y == y then -- skip any NaN defensively
			out:write(name, ",", tostring(i), ",", bits(x), ",", bits(y), "\n")
		end
	end
end
out:close()
io.write("luaprobe done: ", arg[1] or "luaprobe.txt", "\n")
